### ISSUE TITLE
A bug in supernovae

### ISSUE THREAD

**Issue by ilyamandel** [TeamCOMPAS/COMPAS#1149]
We have a bug in SSE [and possibly BSE -- I only checked SSE for now].

Example command: ./COMPAS --mode SSE --initial-mass 6.20
(with COMPAS v02.49.04)

This ultimately triggers a call to ResolveElectronCaptureSupernova(), which should therefore leave behind a OXYGEN_NEON_WHITE_DWARF [checked by putting in print statements].

However, that call only happens via ResolveMassLoss() in Star::EvolveOneTimestep().  And ResolveMassLoss() doesn't actually execute the switch to this new type -- instead, it throws a warning because it doesn't expect the type to change.  So at the end of ResolveMassLoss(), from which ResolveElectronCaptureSupernova() was called, the type is still TPAGB [checked by putting in print statements].

Then, after EvolveOneTimestep() is called from Star::Evolve(), we immediately call UpdateAttributes(0.0. 0.0, true).  By now, however, the star has lost its remaining envelope because it went through a SN, as ResolveElectronCaptureSupernova() set the core mass and the mass to the same value.  Yet it's still a star of the TPAGB type, as its type and class have not yet been updated.   

So when ResolveSupernova() is called from UpdateAttributes, TPAGB::IsSupernova() evaluates to false, and a SN doesn't happen.  As a result, we are left with a COWD, not an ONeWD.

Now, one can quibble whether this should have been a COWD or an ONeWD in reality -- but I think it's quite clear that the code is not doing what was intended, i.e., not setting the star to be an ONeWD as the initial evaluation of ResolveElectronCaptureSupernova() demands.

My preferred solution would be for a star to actually go supernova and change type accordingly when it successfully goes through a supernova function such as ResolveElectronCaptureSupernova().  A supernova is a more "extreme" event than mere mass loss, and so should preempt other behaviour.  However, since the order of calls here is very convoluted, I am rather afraid of how many things I might break if I try to fix this -- so, @jeffriley , could you please take a look?

**Comment by jeffriley** [comment]
I don't really like the idea of switching-in-place with the code the way it is.  Hurley's BSE effectively is a big switch statement for stellar type, and he will change stellar type in one of the cases, but continue executing the remainder of the code in that case - so effectively executing code that is for the previous stellar type...  We kind-of do the same thing, in that we decide we want to switch to a new stellar type, but only do it at the timestep boundary - with the exception (now) that we switch the secondary to a MR for merger products.

I don't really like our current approach (we finish the timestep as though we haven't switched, but we know we should switch...), but I also don't like switching-in-place - because once we switch we are (or could be) executing a function for the previous stellar type, and then we might end up calling a function  that will now be for the new stellar type - we could end up in some weird state.

What I would like to do is allow switch-in-place, but once the switch is done, end the timestep and start a new one - so we always execute a timestep as just one stellar type.  The problem with that - at least I think it may be a problem - is  that we calculate mass loss, mass transfer etc. at the start of the timestep, based on the expected duration of the timestep, and if we end the timestep early what does that mean?  What if when we switch we  have already applied mass loss and mass transfer - do we end the timestep at the switch and assume it was the expected duration?  Maybe we do...  That was one of my worries about "only doing one thing during a timestep".  I can see the attraction, but I think we have to be careful about how we actually implement that. And if we do end a timestep "early", what might we not have done to set up the next timestep?  Maybe nothing, but something we should check

I think I can change the code so that we switch-in-place (i.e. when we decide we should switch, we do, then end the timestep), but I'd prefer to do that after the error handling PR is merged - it will be easier to do with the infrastructure that PR sets up.  And once we have a mechanism for sensibly ending a timestep "early", we can use it for any significant event - not just switching stellar type.

**Comment by ilyamandel** [comment]
@jeffriley -- as you already know, I agree with you, so just publicly confirming this here.  

As you also know, I think of your comment about needing to check what we might have forgotten to do when dealing with the change in stellar type as one more argument in favour of doing one "important" thing per time step.  We want to ease the burden of future development, and that means limiting the range of possible implications that every developer has to think through. :-)

My current thinking is that we should check for dramatic things like SNe at the start of a time step.  If the conditions are satisfied, we should take a size 0 step to take care of these special events.  If not, we take a normal time step.  But I agree that this can be deferred to be part of a broader change to time stepping.

**Comment by ilyamandel** [comment]
This one is still an issue -- hopefully will be solved when we move to doing one important thing at a time...

**Comment by ilyamandel** [comment]
@jeffriley :
Having spent a lot of time on this today, I am afraid this double call to ResolveSupernova() is a much more serious problem than I originally thought.  On the first call, we update the m_Mass variable.  Then we make a call again, and we really do resolve the SN -- but with an updated value of m_Mass, we end up with the wrong SN outcome: wrong final mass, wrong kick, etc.  So this is really serious. :( 
On the plus side, it's only an issue for SSE; I think it was introduced when you attempted to ensure the same behaviour for SSE and BSE.

[Aside for self for future tests -- for the particular example listed above, also check out ./COMPAS --mode SSE --initial-mass 6.20 -n 1 --allow-non-stripped-ECSN ]

**Comment by jeffriley** [comment]
> I think it was introduced when you attempted to ensure the same behaviour for SSE and BSE.

Yes, it was - and if we take it out there will be side effects (COMPAS won't look like SSE anymore...)

**Comment by ilyamandel** [comment]
I believe that SSE and BSE are treated differently *with the current code*.  

Put in a print statement inside the if(IsSupernova()) block in GiantBranch::ResolveSupernova() and run 

./COMPAS --mode SSE --initial-mass 6.20 -n 1 
vs
./COMPAS --mode BSE --initial-mass-1 6.20 --initial-mass-2 1.20 -a 20 -n 1

The first one will print twice.

**Comment by jeffriley** [comment]
@ilyamandel

With COMPAS v03.11.00 I get:

$ ./COMPAS --mode SSE --initial-mass 6.20 -n 1
    
    0: Simulation completed: RandomSeed = 1736930152, Initial Mass = 6.2, Metallicity = 0.0142, Neutron_Star

$ ./COMPAS --mode BSE --initial-mass-1 6.20 --initial-mass-2 1.20 -a 20 -n 1
    
    0: Double White Dwarf formed: (Main_Sequence_>_0.7 -> Oxygen-Neon_White_Dwarf) + (Main_Sequence_>_0.7 -> Carbon-Oxygen_White_Dwarf)

However, if I lower the mass of star 2 I get:

$ ./COMPAS --mode BSE --initial-mass-1 6.20 --initial-mass-2 0.1 -a 20 -n 1
       
    0: Allowed time exceeded: (Main_Sequence_>_0.7 -> Neutron_Star) + (Main_Sequence_<=_0.7 -> Main_Sequence_<=_0.7)

so there must be some interaction happening for the the first BSE run.

I have determined the cause of SSE producing an NS instead of an ONeWD:

The following function was added to the TPAGB class in v03.10.02:

    bool TPAGB::IsSupernova() const {
        if(utils::SNEventType(m_SupernovaDetails.events.current) != SN_EVENT::NONE)
            return true;                                                                            // already labeled as going through a SN right now
        double snMass = CalculateInitialSupernovaMass();                                            // calculate SN initial mass
        return ( utils::Compare(m_COCoreMass, m_GBParams[static_cast<int>(GBP::McSN)]) >=0 && utils::Compare(snMass, OPTIONS->MCBUR1()) >= 0 && utils::Compare(m_COCoreMass, m_Mass) < 0 );
        // no supernova if CO core mass is too low or helium core mass is too low at base of AGB or the envelope has already been removed
    }

The first conditional should be removed.  The `IsSupernova()` functions should return `TRUE` if the attributes of the star satisfy the criteria for the star contining evolution as a supernova (e.g. mass exceeds some threshold).  `IsSupernova()` should not return `TRUE` on the basis of the `SN_EVENT` flag being set - that indicates that the star underwent a SN in this timestep, not that it needs to be evolved as a supernova about to happen.

Removing the conditional:

        if(utils::SNEventType(m_SupernovaDetails.events.current) != SN_EVENT::NONE)
            return true;                                                                            // already labeled as going through a SN right now

from `TPAGB::IsSupernova()` prevents the second call to GiantBranch::ResolveSupernova() executing the body of the function, and results in:

$ ./COMPAS --mode SSE --initial-mass 6.20 -n 1

    0: Simulation completed: RandomSeed = 1736930936, Initial Mass = 6.2, Metallicity = 0.0142, Oxygen-Neon_White_Dwarf


$ ./COMPAS --mode BSE --initial-mass-1 6.20 --initial-mass-2 1.20 -a 20 -n 1

    0: Double White Dwarf formed: (Main_Sequence_>_0.7 -> Oxygen-Neon_White_Dwarf) + (Main_Sequence_>_0.7 -> Carbon-Oxygen_White_Dwarf)


$ ./COMPAS --mode BSE --initial-mass-1 6.20 --initial-mass-2 0.1 -a 20 -n 1

    0: Allowed time exceeded: (Main_Sequence_>_0.7 -> Neutron_Star) + (Main_Sequence_<=_0.7 -> Main_Sequence_<=_0.7)


Does this change resolve the reason this issue was originally opened?  Or is there still a problem because SSE and BSE don't give the same results?

**Comment by jeffriley** [comment]
@ilyamandel 

I ran this command (with the conditional removed as described above):

    ./COMPAS --mode SSE --random-seed 0 --initial-mass 6.20 -n 1 --detailed-output --logfile-type csv

which resulted in:

    0: Simulation completed: RandomSeed = 0, Initial Mass = 6.2, Metallicity = 0.0142, Oxygen-Neon_White_Dwarf
    
and harvested the timesteps from the detailed output file.  I then ran this command (again, with the conditional removed as described above), with the timesteps harvested from the SSE run:

    ./COMPAS --mode BSE --random-seed 0 --initial-mass-1 6.20 --initial-mass-2 0.1 -a 1E20 -n 1 --timesteps-filename timesteps.txt --detailed-output --logfile-type csv

which resulted in:

    0: User-provided timesteps exhausted: (Main_Sequence_>_0.7 -> Thermally_Pulsing_Asymptotic_Giant_Branch) + (Main_Sequence_<=_0.7 -> Main_Sequence_<=_0.7)

The detailed outputs from each run show that the evolution of the primary is the same up until the last timestep - in SSE the star evolves to an OnEWD, but in BSE it remains a TPAGB star.  I haven't determined why - I probably won't have time before I leave for Italy - but it could just be that we have made changes, especially to the BSE code, since I did the work to ensure SSE and BSE produced the same results with the same timesteps.  I can follow-up when I'm back from Italy if you think it is necessary.

Do you want a PR opened to remove the conditional in `TPAGB::IsSupernova()` (as described above)?

**Comment by jeffriley** [comment]
Detailed output files (SSE and BSE) from runs described above:

[SSE_Detailed_Output_0.csv](https://github.com/user-attachments/files/18429819/SSE_Detailed_Output_0.csv)
[BSE_Detailed_Output_0.csv](https://github.com/user-attachments/files/18429833/BSE_Detailed_Output_0.csv)

**Comment by ilyamandel** [comment]
@jeffriley -- thank you very much for making this fix!
I want to return to the original issue and check further before closing this one, but don't have time to do that right now.  I will, however, change the labels on this one given your PR.
